src/PIL/DdsImagePlugin.py

```diff
@@ -148,15 +147,38 @@ def _open(self):
             self.tile = [("raw", (0, 0) + self.size, 0, (self.mode, 0, 1))]
         elif pfflags & DDPF_RGB:
             # Texture contains uncompressed RGB data
-            masks = {mask: ["R", "G", "B", "A"][i] for i, mask in enumerate(masks)}
-            rawmode = ""
-            if pfflags & DDPF_ALPHAPIXELS:
-                rawmode += masks[0xFF000000]
-            else:
+            if not pfflags & DDPF_ALPHAPIXELS:
                 self._mode = "RGB"
-            rawmode += masks[0xFF0000] + masks[0xFF00] + masks[0xFF]
 
-            self.tile = [("raw", (0, 0) + self.size, 0, (rawmode[::-1], 0, 1))]
+            # Each channel is derived using a mask of the pixel values
+            masks = {
+                mask: "RGB"[i]
+                for i, mask in enumerate(struct.unpack("<3I", header.read(12)))
+            }
+            if self.mode == "RGB" and all(
+                mask in (0b0000000000011111, 0b0000001111100000, 0b0111110000000000)
```

Comment:
This did become less clear to read, you're right. Wonder if these and the `0xFF0000` masks from below could benefit from becoming constants á la (I didn't double check whether I have RGB/BGR right here):

```
R_MASK_555 = 0b0000000000011111
G_MASK_555 = 0b000001111100000
B_MASK_555 = 0b111110000000000
R_MASK_888 = 0x000000FF
G_MASK_888 = 0x0000FF00
B_MASK_888 = 0x00FF0000
A_MASK_888 = 0xFF000000
```